pkg/controllers/federatedhpa/federatedhpa_controller.go

```diff
@@ -705,118 +735,163 @@ func (c *FederatedHPAController) computeReplicasForMetric(ctx context.Context, h
-			return 0, time.Time{}, "", condition, fmt.Errorf("failed to get %s object metric: %v", metricSpec.Object.Metric.Name, err)
+			resp.condition = c.getUnableComputeReplicaCountCondition(req.hpa, "FailedGetObjectMetric", err)
+			resp.timestampProposal = time.Time{}
+			return resp, fmt.Errorf("failed to get %s object metric: %v", req.metricSpec.Object.Metric.Name, err)
 		}
-		*status = autoscalingv2.MetricStatus{
+		resp.status = &autoscalingv2.MetricStatus{
 			Type: autoscalingv2.ObjectMetricSourceType,
 			Object: &autoscalingv2.ObjectMetricStatus{
 				Metric: autoscalingv2.MetricIdentifier{
-					Name:     metricSpec.Object.Metric.Name,
-					Selector: metricSpec.Object.Metric.Selector,
+					Name:     req.metricSpec.Object.Metric.Name,
+					Selector: req.metricSpec.Object.Metric.Selector,
 				},
 				Current: autoscalingv2.MetricValueStatus{
 					AverageValue: resource.NewMilliQuantity(usageProposal, resource.DecimalSI),
 				},
 			},
 		}
-		return replicaCountProposal, timestampProposal, fmt.Sprintf("external metric %s(%+v)", metricSpec.Object.Metric.Name, metricSpec.Object.Metric.Selector), autoscalingv2.HorizontalPodAutoscalerCondition{}, nil
+		resp.timestampProposal = timestampProposal
+		resp.condition = autoscalingv2.HorizontalPodAutoscalerCondition{}
+		resp.replicaCountProposal = replicaCountProposal
+		resp.metricNameProposal = fmt.Sprintf("external metric %s(%+v)", req.metricSpec.Object.Metric.Name, req.metricSpec.Object.Metric.Selector)
+		return resp, nil
 	}
 	errMsg := "invalid object metric source: neither a value target nor an average value target was set"
 	err = fmt.Errorf(errMsg)
-	condition = c.getUnableComputeReplicaCountCondition(hpa, "FailedGetObjectMetric", err)
-	return 0, time.Time{}, "", condition, err
+	resp.condition = c.getUnableComputeReplicaCountCondition(req.hpa, "FailedGetObjectMetric", err)
+	resp.timestampProposal = time.Time{}
+	return resp, err
 }
 
 // computeStatusForPodsMetric computes the desired number of replicas for the specified metric of type PodsMetricSourceType.
-func (c *FederatedHPAController) computeStatusForPodsMetric(currentReplicas int32, metricSpec autoscalingv2.MetricSpec, hpa *autoscalingv1alpha1.FederatedHPA, selector labels.Selector, status *autoscalingv2.MetricStatus, metricSelector labels.Selector, podList []*corev1.Pod, calibration float64) (replicaCountProposal int32, timestampProposal time.Time, metricNameProposal string, condition autoscalingv2.HorizontalPodAutoscalerCondition, err error) {
-	replicaCountProposal, usageProposal, timestampProposal, err := c.ReplicaCalc.GetMetricReplicas(currentReplicas, metricSpec.Pods.Target.AverageValue.MilliValue(), metricSpec.Pods.Metric.Name, hpa.Namespace, selector, metricSelector, podList, calibration)
+func (c *FederatedHPAController) computeStatusForPodsMetric(req *hpaReqParams) (resp hpaResp, err error) {
+	replicaCountProposal, usageProposal, timestampProposal, err := c.ReplicaCalc.GetMetricReplicas(
```

Comment:
It seems that we can only utilize `req` as the parameter.

Reply:
I agree with @jwcesign. There is also the problem of too many function params in `ReplicaCalculator`. We'd better unify it.

Reply:
how about current one?